app/user/user_main.c

```diff
@@ -187,60 +184,73 @@ void user_rf_pre_init(void)
- *                A : rf cal
- *                B : rf init data
- *                C : sdk parameters
- * Parameters   : none
- * Returns      : rf cal sector
-*******************************************************************************/
-uint32
-user_rf_cal_sector_set(void)
+#define EAGLE_FLASH_BIN_ADDR      (SYSTEM_PARTITION_CUSTOMER_BEGIN + 1)
+#define EAGLE_IROM0TEXT_BIN_ADDR  (SYSTEM_PARTITION_CUSTOMER_BEGIN + 2)
+static const partition_item_t pt_template[] ICACHE_RODATA_ATTR STORE_ATTR = {
+    { EAGLE_FLASH_BIN_ADDR,     0x00000, 0x10000},
+    { EAGLE_IROM0TEXT_BIN_ADDR, 0x10000, 0x60000},
+    { SYSTEM_PARTITION_RF_CAL, 0, 0x1000},
+    { SYSTEM_PARTITION_PHY_DATA, 0, 0x1000},
+    { SYSTEM_PARTITION_SYSTEM_PARAMETER, 0, 0x3000},
+};
+
+void ICACHE_FLASH_ATTR user_pre_init(void)
 {
-    enum flash_size_map size_map = system_get_flash_size_map();
-    uint32 rf_cal_sec = 0;
+    partition_item_t pt[sizeof(pt_template)/sizeof(pt_template[0])];
+    memcpy(pt, pt_template, sizeof(pt_template));
 
+    enum flash_size_map size_map = system_get_flash_size_map();
     switch (size_map) {
-        case FLASH_SIZE_4M_MAP_256_256:
-            rf_cal_sec = 128 - 5;
+        case FLASH_SIZE_8M_MAP_512_512: // 2
+            pt[2].addr = 0xfb000;
+            pt[3].addr = 0xfc000;
+            pt[4].addr = 0xfd000;
             break;
-
-        case FLASH_SIZE_8M_MAP_512_512:
-            rf_cal_sec = 256 - 5;
+        case FLASH_SIZE_16M_MAP_512_512: // 3
+        case FLASH_SIZE_16M_MAP_1024_1024: // 5
+            pt[2].addr = 0x1fb000;
+            pt[3].addr = 0x1fc000;
+            pt[4].addr = 0x1fd000;
             break;
-
-        case FLASH_SIZE_16M_MAP_512_512:
-        case FLASH_SIZE_16M_MAP_1024_1024:
-            rf_cal_sec = 512 - 5;
+        case FLASH_SIZE_32M_MAP_512_512: // 4
+        case FLASH_SIZE_32M_MAP_1024_1024: // 6
+        case FLASH_SIZE_32M_MAP_2048_2048: // 7
+            pt[2].addr = 0x3fb000;
+            pt[3].addr = 0x3fc000;
+            pt[4].addr = 0x3fd000;
             break;
-
-        case FLASH_SIZE_32M_MAP_512_512:
-        case FLASH_SIZE_32M_MAP_1024_1024:
-        case FLASH_SIZE_32M_MAP_2048_2048:
-            rf_cal_sec = 1024 - 5;
+        case FLASH_SIZE_64M_MAP_1024_1024: // 8
+            // XXX No examples given in docs; best guess
+            pt[2].addr = 0x7fb000;
+            pt[3].addr = 0x7fc000;
+            pt[4].addr = 0x7fd000;
             break;
-
-        case FLASH_SIZE_64M_MAP_1024_1024:
-            rf_cal_sec = 2048 - 5;
-            break;
-
-        case FLASH_SIZE_128M_MAP_1024_1024:
-            rf_cal_sec = 4096 - 5;
+        case FLASH_SIZE_128M_MAP_1024_1024:  // 9
+            // XXX No examples given in docs; best guess
+            pt[2].addr = 0xffb000;
+            pt[3].addr = 0xffc000;
+            pt[4].addr = 0xffd000;
             break;
 
+        case FLASH_SIZE_4M_MAP_256_256: /* 0: Too small */
```

Comment:
I think you should decode this case as well. Even if a 512kB flash might be too small, this `size_map` is the default at some points in the tool chain in the absence of a specific value.
Blocking the firmware startup in this case prevents error/diagnostic messages further down the startup code.